File: crates/event-log/src/artifacts/copy.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use super::records::{advance, invalid, read_record, require_session};
use crate::StoreError;
use maka_runtime::{
    artifact::ArtifactSource,
    attachment::{AttachmentRef, MAX_ATTACHMENT_BYTES, StorageRef},
};
use sqlx::SqliteConnection;

/// Copy an explicitly authorized attachment without materializing its BLOB in the Host.
/// The caller commits this together with the message that owns the destination.
pub(crate) async fn copy_in_transaction(
    tx: &mut SqliteConnection,
    source: &AttachmentRef,
    destination: &AttachmentRef,
    turn: &str,
    created_at: u64,
) -> Result<(), StoreError> {
    let StorageRef::SessionFile {
        session_id: source_session,
        relative_path: source_id,
    } = &source.storage_ref
    else {
        return Err(invalid("Attachment source is not a Session Artifact"));
    };
    let StorageRef::SessionFile {
        session_id: target_session,
        relative_path: target_id,
    } = &destination.storage_ref
    else {
        return Err(invalid("Attachment destination is not a Session Artifact"));
    };
    let (mut artifact, digest) = read_record(tx, source_session, source_id)
        .await?
        .ok_or_else(|| invalid("Delegated attachment source no longer exists"))?;
    artifact.validate_attachment(source).map_err(invalid)?;
    if artifact.size_bytes > MAX_ATTACHMENT_BYTES
        || artifact.source == ArtifactSource::UserUpload
            && artifact.summary.as_ref() != Some(&digest)
    {
        return Err(StoreError::ArtifactConflict);
    }
    require_session(tx, target_session).await?;
    artifact.session_id = target_session.clone();
    artifact.id = target_id.clone();
    artifact.turn_id = turn.into();
    artifact.created_at = created_at;
    artifact.validate().map_err(invalid)?;
    artifact.validate_attachment(destination).map_err(invalid)?;
    let encoded = serde_json::to_string(&artifact)?;
    if encoded.len() > super::MAX_RECORD_BYTES {
        return Err(invalid("Copied attachment metadata exceeds limit"));
    }
    if let Some((existing, target_digest)) = read_record(tx, target_session, target_id).await? {
        if existing != artifact || target_digest != digest {
            return Err(StoreError::ArtifactConflict);
        }
        let same: bool = sqlx::query_scalar(
            "SELECT EXISTS(SELECT 1 FROM artifacts s JOIN artifacts t
             ON t.session_id = ?3 AND t.id = ?4 WHERE s.session_id = ?1 AND s.id = ?2
             AND length(s.payload) = ?5 AND s.payload = t.payload)",
        )
        .bind(source_session)
        .bind(source_id)
        .bind(target_session)
        .bind(target_id)
        .bind(artifact.size_bytes as i64)
        .fetch_one(tx)
        .await?;
        return if same {
            Ok(())
        } else {
            Err(StoreError::ArtifactConflict)
        };
    }
    let inserted = sqlx::query(
        "INSERT INTO artifacts (session_id, id, created_at, record_json, content_sha256, payload)
         SELECT ?3, ?4, ?5, ?6, content_sha256, payload FROM artifacts
         WHERE session_id = ?1 AND id = ?2 AND length(payload) = ?7",
    )
    .bind(source_session)
    .bind(source_id)
    .bind(target_session)
    .bind(target_id)
    .bind(created_at as i64)
    .bind(encoded)
    .bind(artifact.size_bytes as i64)
    .execute(&mut *tx)
    .await?
    .rows_affected();
    if inserted != 1 {
        return Err(StoreError::ArtifactConflict);
    }
    advance(tx, target_session).await
}

impl crate::EventLog {
    /// Idempotent immutable copy, authorized by Host at both endpoints. Payloads
    /// stay inside SQLite; even a 50 MiB attachment never crosses a V8 heap.
    pub async fn copy_plugin_attachment(
        &self,
        namespace: &maka_plugins::storage::Namespace,
        request: maka_plugins::execution::CopyAttachment,
    ) -> Result<AttachmentRef, StoreError> {
        use sha2::{Digest, Sha256};
        use sqlx::Connection;
        self.validate_root()?;
        request
            .validate()
            .map_err(|error| invalid(&error.to_string()))?;
        let identity = serde_json::to_vec(&(namespace.package(), namespace.scope(), &request))?;
        let id = format!("attachment-{:x}", Sha256::digest(identity));
        let mut destination = request.attachment.clone();
        destination.storage_ref = StorageRef::SessionFile {
            session_id: request.target_session_id.clone(),
            relative_path: id.clone(),
        };
        self.connection
            .run(move |connection| {
                Box::pin(async move {
                    let mut tx = connection.begin_with("BEGIN IMMEDIATE").await?;
                    let created_at =
                        match read_record(&mut tx, &request.target_session_id, &id).await? {
                            Some((artifact, _)) => artifact.created_at,
                            None => std::time::SystemTime::now()
                                .duration_since(std::time::UNIX_EPOCH)
                                .map_err(|error| invalid(&error.to_string()))?
                                .as_millis()
                                .try_into()
                                .map_err(|_| invalid("clock overflow"))?,
                        };
                    copy_in_transaction(
                        &mut tx,
                        &request.attachment,
                        &destination,
                        &id,
                        created_at,
                    )
                    .await?;
                    tx.commit().await.map_err(StoreError::CommitUnknown)?;
                    Ok(destination)
                })
            })
            .await
    }
}

File: crates/scheduler/src/client/copy.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type { ClientLocale } from '@maka-agent/plugin-sdk/client';

const en = {
  authorize: 'Allow scheduled background work',
  authorizationDenied: 'Background authorization was not granted',
  deleteTitle: 'Delete scheduled task?',
  deleteDescription: 'Already accepted executions will not be cancelled.',
  noModel: 'This Session has no model',
  accessTitle: 'Scheduled background access',
  allowResume: 'Allow resuming this Session',
  allowCreate: 'Allow new Sessions in this workspace',
  activeTasks: 'Active scheduled tasks',
};

export const copy = {
  en,
  'zh-CN': {
    authorize: '允许定时任务在后台执行',
    authorizationDenied: '未获得后台授权',
    deleteTitle: '删除定时任务？',
    deleteDescription: '已接受的执行不会被取消。',
    noModel: '此会话没有模型',
    accessTitle: '定时任务后台权限',
    allowResume: '允许继续此会话',
    allowCreate: '允许在此工作区创建会话',
    activeTasks: '活动定时任务',
  },
  'zh-TW': {
    authorize: '允許排程工作在背景執行',
    authorizationDenied: '未取得背景授權',
    deleteTitle: '刪除排程工作？',
    deleteDescription: '已接受的執行不會被取消。',
    noModel: '此對話沒有模型',
    accessTitle: '排程工作背景權限',
    allowResume: '允許繼續此對話',
    allowCreate: '允許在此工作區建立對話',
    activeTasks: '執行中的排程工作',
  },
} satisfies Record<ClientLocale, typeof en>;
